udfs/remote_udfs/nlp/README.md

```diff
@@ -0,0 +1,121 @@
+```
+
+Within properties you'll find *serviceAccountId* which will have the service ID you'll need in a subsequent step.
+
+### Deploying your Cloud Function
+
+[More information about deploying your cloud function can be found here.](https://cloud.google.com/functions/docs/deploy)
+
+Snippet provided below for brevity.  
+It’s recommended that you keep the default authentication instead of allowing unauthenticated invocation of your Cloud Function or Cloud Run service.  
+We use gen1 Cloud Functions here for the simple demo purposes; however, gen2 Cloud Functions are recommended. 
+
+* project - the project the cloud function is deployed to 
+* runtime - this was defaulted to python39 but can be changed as required 
+```
+gcloud functions deploy sampleCF \
+--project=$PROJECT --runtime=python39 --entry-point=remote_vertex_ai --source=call_nlp --trigger-http
+```
+
+### Granting the service account invoker permissions on the functions
+
+Grant the service account obtained above permissions to invoke the functions.
+* project - the project where your external connection was created
+* cf_name - the name of your cloud function 
+* service_account - the service account obtained above
+
+**_NOTE:_** For the below command you will need to have jq installed. 
+
+```
+SERVICE_ACCOUNT=$(bq show --connection --project_id=$PROJECT --location=$LOCATION remote_connection | jq '.cloudResource.serviceAccountId' | tr -d '"')
+gcloud --project=$PROJECT functions add-iam-policy-binding sampleCF --member=serviceAccount:$SERVICE_ACCOUNT --role=roles/cloudfunctions.invoker
+```
+
+### Obtaining the full Cloud Function endpoint
+
+[More information about the gcloud functions describe can be found here.](https://cloud.google.com/sdk/gcloud/reference/functions/describe)
+* cf_name - the name of the cloud function
+```
+gcloud functions describe sampleCF
+```
+
+You will need the full URL under the httpsTrigger section.
+It should look something like this:
+```
+https://<location>-<project-id>.cloudfunctions.net/<cf-name>
+```
+
+### Creating your BigQuery UDF on BigQuery
+
+You can choose to do this many ways.  
+* [creating your bq function](/remote_udfs/nlp/create_bq_function.sh) 
+
+The input parameters for the helper script above need to be in order:
+* project - the project you deployed to 
+* dataset - the BigQuery dataset you want to deploy to 
+* location - the location where your BigQuery dataset is
+* endpoint - the full endpoint you obtained above
+
+```
+ENDPOINT=$(gcloud functions describe $CF_NAME --format="value(httpsTrigger.url)")
+
+sh create_bq_function.sh $PROJECT $DATASET $LOCATION $PROJECT.$LOCATION.sampleCF $ENDPOINT
+```
+
+### Running it on BigQuery
+You should now be able to run the remote UDF on BigQuery.
+
+Try it in your BigQuery console. 
+```
+select %your_project_id%.%your_dataset_id%.%function_name%("This is really awesome!");
+```
+
+It should return a positive float value (as in greater than 0) as a response.
+
```

Comment:
Would be good to have a "putting it all together" section at the end where you have all the steps in one script for user to easily run:

```
PROJECT=your_project_id
LOCATION=your_bigquery_dataset_location
DATASET=your_bigquery_dataset_id

bq mk --connection \
    --display_name=remote_connection \
    --connection_type=CLOUD_RESOURCE \
    --project_id="${PROJECT}" \
    --location="${LOCATION}" \
    remote_connection

gcloud functions deploy sampleCF \
    --project="${PROJECT}" \
    --runtime=python39 \
    --entry-point=remote_vertex_ai \
    --source=call_nlp \
    --trigger-http

SERVICE_ACCOUNT=$(bq show --connection --project_id="${PROJECT}" --location="${LOCATION}" remote_connection | jq '.cloudResource.serviceAccountId' -r)
gcloud functions add-iam-policy-binding sampleCF \
    --project="${PROJECT}" \
    --member=serviceAccount:"${SERVICE_ACCOUNT}" \
    --role=roles/cloudfunctions.invoker

ENDPOINT=$(gcloud functions describe sampleCF --format="value(httpsTrigger.url)")
NLP_QUERY="""
CREATE OR REPLACE  FUNCTION \`${PROJECT}.${DATASET}.call_nlp\` (x STRING)
RETURNS STRING
REMOTE WITH CONNECTION \`${PROJECT}.${LOCATION}.remote_connection\`
OPTIONS(
  endpoint = '${ENDPOINT}',
  user_defined_context = [(\"mode\",\"call_nlp\")]
);
"""

bq query \
    --location="$LOCATION" \
    --use_legacy_sql=false \
    "${NLP_QUERY}"

```